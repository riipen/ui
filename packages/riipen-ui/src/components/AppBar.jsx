import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";

class AppBar extends React.Component {
  static propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.object,

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     */
    color: PropTypes.oneOf(["primary", "secondary", "white", "black"]),

    /**
     * The positioning type. The behavior of the different options is described
     * [in the MDN web docs](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Positioning).
     * Note: `sticky` is not universally supported and will fall back to `static` when unavailable.
     */
    position: PropTypes.oneOf([
      "absolute",
      "fixed",
      "relative",
      "static",
      "sticky"
    ])
  };

  static defaultProps = {
    color: "primary",
    position: "fixed"
  };

  static contextType = ThemeContext;

  render() {
    const { children, classes, color, position } = this.props;

    const theme = this.context;

    const className = clsx(color, position, classes);

    return (
      <React.Fragment>
        <div className={className}>{children}</div>
        <style jsx>{`
          div {
            align-items: center;
            box-shadow: ${theme.shadows[3]};
            box-sizing: border-box;
            display: flex;
            padding: ${theme.spacing(3)}px;
            width: 100%;
            z-index: ${theme.zIndex.low};
          }

          .fixed {
            left: auto;
            position: fixed;
            right: 0;
            top: 0;
          }
          .absolute {
            left: auto;
            position: absolute;
            right: 0;
            top: 0;
          }
          .stickey {
            left: auto;
            position: sticky;
            right: 0;
            top: 0;
          }
          .static {
            position: static;
          }
          .relative {
            position: relative;
          }

          .primary {
            background-color: ${theme.palette.primary.main};
            color: ${theme.palette.primary.contrast};
          }
          .secondary {
            background-color: ${theme.palette.secondary.main};
            color: ${theme.palette.secondary.contrast};
          }
          .white {
            background-color: ${theme.palette.common.white};
            color: ${theme.palette.common.black};
          }
          .black {
            background-color: ${theme.palette.common.black};
            color: ${theme.palette.common.white};
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default AppBar;
