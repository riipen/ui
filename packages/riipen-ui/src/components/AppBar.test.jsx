import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import AppBar from "./AppBar";

describe("<AppBar>", () => {
  it("renders without errors", () => {
    let error;
    try {
      mount(<AppBar />);
    } catch (theError) {
      error = theError;
    }
    expect(error).toEqual(undefined);
  });

  it("sets correct default props", () => {
    const defaultProps = new AppBar().type.defaultProps;

    const wrapper = mount(<AppBar />);

    const component = wrapper.find("AppBar").childAt(0);
    expect(component.hasClass(defaultProps.color)).toEqual(true);
    expect(component.hasClass(defaultProps.position)).toEqual(true);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("children prop", () => {
    it("displays custom label from child", () => {
      const label = "A label given as a child";
      const child = <div>{label}</div>;

      const wrapper = mount(<AppBar>{child}</AppBar>);

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
      expect(wrapper.text(".label")).toEqual(label);
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classVariant = ["newClass"];

      const wrapper = mount(<AppBar classes={classVariant} />);

      expect(
        wrapper
          .find("AppBar")
          .childAt(0)
          .hasClass(classVariant[0])
      ).toEqual(true);
    });
  });

  describe("color prop", () => {
    it("sets color class name with valid custom colour", () => {
      const colorVariant = "secondary";

      const wrapper = mount(<AppBar color={colorVariant} />);

      expect(wrapper.find("AppBar").props().color).toEqual(colorVariant);
    });

    it("sets color class name with invalid colour", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<AppBar color="orange" />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("position prop", () => {
    it("sets position class name with valid custom position", () => {
      const positionVariant = "absolute";

      const wrapper = mount(<AppBar color={positionVariant} />);

      expect(wrapper.find("AppBar").props().color).toEqual(positionVariant);
    });

    it("sets position class name with invalid position", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<AppBar position="upright" />);

      expect(errors).toHaveBeenCalled();
    });
  });
});
