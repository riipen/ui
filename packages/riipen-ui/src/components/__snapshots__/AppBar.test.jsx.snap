// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppBar> renders correct snapshot 1`] = `
<withClasses(AppBar)>
  <AppBar
    classes={
      Array [
        "riipen",
        "riipen-appbar",
      ]
    }
    color="primary"
    position="fixed"
  >
    <div
      className="jsx-138222299 riipen riipen-appbar primary fixed"
    />
    <JSXStyle
      dynamic={
        Array [
          "0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12)",
          15,
          2,
          "#1f71ae",
          "#fff",
          "#f58632",
          "#fff",
          "#fff",
          "#000",
          "#000",
          "#fff",
        ]
      }
      id="4181421979"
    />
  </AppBar>
</withClasses(AppBar)>
`;
